### DISCUSSION TITLE
Proposal: Add support for typedef (like in C++)

### DISCUSSION THREAD

**Discussion by aaronfranke** [dotnet/csharplang#1300] [Language Ideas] [answered]
Currently, in C#, there is no `typedef` like in C++. My use case for `typedef` is writing math methods with simultaneous support for `float` and `double` (an abstract type, via `typedef`, would be used in place of either type). Or when creating a descriptive short-hand for types without objects.

Would it be possible to add this functionality to the C# language spec?

**Comment by benaadams** [comment]
You can do aliasing currently for types based on defines with use `using ... = ...` with `#if` to achieve the `float`/`double`
```csharp
#if REAL_IS_DOUBLE
using real = System.Double;
#else
using real = System.Single;
#endif
```

```csharp
real Add(real a, real b)
{
    return a + b;
}
```

Or a native int type (assuming you set `#define BIT64` for a 64 bit build)
```csharp
#if BIT64
using nint = System.Int64;
#else
using nint = System.Int32;
#endif
```

**Comment by stakx** [comment]
... and you have to repeat the alias definition in each source file where you want to use the alias.

**Reply by DavidArno** [reply]
@stakx,

This is no longer true. These days the following is only needed in one source file for the entire assembly:
```cs
#if BIT64
global using nint = System.Int64;
#else
global using nint = System.Int32;
#endif
```

**Reply by BinToss** [reply]
@DavidArno Since when (which lang version) was this a thing?

EDIT: C# 10 and .NET 6

**Reply by miyu** [reply]
This doesn't scale to larger multi-project solutions; if you're compiling a program variant and this typedef is returned or ingested in your API, you'd then typically prefer your callers can also use your type alias (e.g. so that they can nint.MaxValue in your example)

**Reply by bernd5** [reply]
Those type aliases are lowered away completely...

**Comment by MkazemAkhgary** [comment]
>... and you have to repeat the alias definition in each source file where you want to use the alias.

I don't think it would be that hard for compiler to treat it as public, yet its not supported.

**Comment by bondsbw** [comment]
@MkazemAkhgary `using` is scoped to the current file, whether it be for importing namespaces, class aliases, or importing static methods.

I would like a global class alias but it would probably be better to use a different keyword.

**Comment by aaronfranke** [comment]
> it would probably be better to use a different keyword.

How about `typedef`? This would be most familiar to people coming from C++.

**Comment by aluanhaddad** [comment]
@aaronfranke `typedef` in C++ has been superseded by the enhanced `using` alias introduced in C++11 so the familiarity argument doesn't track. If anything `using` should be enhanced in C#. I also don't see how it is an abstract type. It is just an alias that you can optionally use.

**Comment by aaronfranke** [comment]
It is useful as an abstract type because you can have situations where a data type can be either A or B, but in-file you just refer to it as type X. A single line can then be changed to make it become A or B.

**Comment by aluanhaddad** [comment]
@aaronfranke 
> It is useful as an abstract type because you can have situations where a data type can be either A or B, but in-file you just refer to it as type X. A single line can then be changed to make it become A or B.

I don't think this is a great way to factor code. I have no doubt it works well enough, but it is not expressive. A new language feature should enable clearer expression of ideas.

**Comment by contextfree** [comment]
It seems like it should be possible at least to have assembly-scoped `using`. "`internal using`"?

For me the use case would be more about wanting to avoid the syntactic overhead of putting IBlahBlahBlah<SomeOtherLongTypeName> everywhere, but not wanting to require clients to use a specific wrapper type.

**Comment by aluanhaddad** [comment]
@contextfree that's an excellent choice of syntax conveying the semantics perfectly. There's also a proposal, #1180, for namespace level method declarations. Such a feature would I think cover that proposal, which I quite like, as well:

```c#
internal using static Namespace.Utilities;

namespace Namespace
{
    static class Utilities
    {
        public static void WriteLine(object value) => Console.WriteLine(value);
    }
}
```
So while I dislike the example use case of this proposal, your suggestion would make it generalizable to much more interesting scenarios.

**Comment by Danielku15** [comment]
To extend this proposal with a more modern approach than typedefs:

Haxe, a language for cross-compilation to other languages, has a concept called [abstract types] (https://haxe.org/manual/types-abstract.html) beside the typedefs . 

It allows you to define a type which is evaluated to its underlying type during compile time. For this type you can define custom operators, methods. If an abstract type cannot have additional members as the storage will be the defined underlying type, but it can have methods, properties and operators. Haxe also supports inlining of methods which might any overhead. If you do not inline the methods, there will be a class holding the methods as statics. This way you could even implement a "int" that can only hold even numbers.

I personally do not like the fact that they call them "abstract types" as `abstract` means something else in C# and Java. But by using `typedef` as keyword we could build a even more powerful system than recycling maybe the C++ style typedef approach. 

```csharp
typedef Color : int 
{
    public inline Color(int i) 
    {
        this = i;
    }
    
    public inline Color(Color c)
    {
        this = (int)c;
    }
    
    public inline byte Alpha => (this >> 24) & 0xFF;
    public inline byte Red => (this >> 16) & 0xFF;
    public inline byte Green => (this >> 8) & 0xFF;
    public inline byte Blue => this & 0xFF;
    
    public static inline Color FromArgb(byte a, byte r, byte g, byte b)
    {
        return new Color((a << 24) | (r << 16) | (g << 8) | b);
    }
    
    public static inline implicit operator int(Color c) 
    {
        return c;
    }
    
    public static inline implicit operator Color(int i) 
    {
        return new Color(i);
    }
}

typedef GameNumber 
#if BIT64
        : double
#else 
        : float
#endif
{
    public void DoSomething(GameEngine engine)
    {
        engine.PutToPool(this); // GameEngine.PutToPool(float f) or GameEngine.PutToPool(double d)
    }
    public static inline explicit operator long(Real c); // auto-operator via underlying type
    public static inline implicit operator Real(long i); // auto-operator via underlying type
}
```

There would be definitly some details to clarify like: 

* Auto exposing of underlying type methods
* Inlining strategy (not sure if the C# compiler already can do that well)
* Exposing of typedefs outside assemblies
* Refection/Dynamic support

This would also allow a bit more strict typing within your library.

**Comment by CyrusNajmabadi** [comment]
I'm not sure what that buys you on top of what you can already do today (just wrapping your value in a struct that has whatever API you want to expose).

**Comment by Danielku15** [comment]
The benefit in this case is that on compile time the type is reduced to the underlying type. This means it can also be a class or interface that you wrap in this case. When you wrap it into a struct you will have a lot of wrapping/unwrapping ongoing. Plus: You will have a new type in your assembly and public interface even though you just wanted to have a wrapper with some more meaning. My proposal is more an extension of a typedef too maybe allow more logic. Also a simple typedef could be rebuild with a struct/class + implicit cast operators but in the end that's not the goal. 

I saw especially in cross-platform compilation cases such types can be really useful especially if you have platform specific implementations. 

The bullet point with "exposing outside assemblies" is maybe misleading. By this I did not mean directly that the typedef is exposed and embedded in the assembly, but rather **if** you have typedefs in your contract, you might need to consider that the conversion operator of the passed value is invoked at the beginning of the method. If you have a `EvenInt` type that ensures only even numbers, and somebody passes an odd integer to your method it might need to call your conversion function. Also the bullet point on reflection and dynamic might just end in the decision that typedefs will provide the functionality of the underlying type and additional members are lost.

**Comment by jnm2** [comment]
> This means it can also be a class or interface that you wrap in this case

I missed your point the first few times I read this line. Meaning, you want to be able to pass your typedef type to things expecting ISomeInterface or SomeBaseClass?

**Comment by Danielku15** [comment]
> Meaning, you want to be able to pass your typedef type to things expecting ISomeInterface or SomeBaseClass

Yes, you can basically wrap anything in the typedef, a class, interface, enum or struct (by this also standard data types like string). Just like in C++ the typedef becomes an alias for the original type. 

The simple version of a typedef could be `typedef MyFloat : float;` In this case it is a simple alias. Variables of MyFloat can get float values assigned or be used anywhere where a float is needed. Also it provides all methods/properties etc. forwarded to the original type. A classical typedef or name-alias. 

The extended version brings a bit more strict typing into the game. At compile time the alias also gets replaced with the underlying type. But additionally you can define specific operators, methods and maybe properties. The code of those might either be inlined for performance reasons or be placed in a static class with extension methods. It covers the use case that the typedef is more than a simple alias for convenience but really a entity in your assembly it has some more semantics and should follow some rules.

**Comment by CyrusNajmabadi** [comment]
I'm not sure what that buys you on top of what you can already do today (just wrapping your value in a struct that has whatever API you want to expose). :)

**Comment by CyrusNajmabadi** [comment]
> The benefit in this case is that on compile time the type is reduced to the underlying type. This means it can also be a class or interface that you wrap in this case. 

Sure, that could also be the case when you wrap with a struct.

> When you wrap it into a struct you will have a lot of wrapping/unwrapping ongoing. 

Wait, why?  The runtime is very smart about this.  A struct that wraps a single value has no overhead.  And because there is no inheritance/virtual/interfaces going on, practically all operations punch-through to the underlying value.

**Comment by CyrusNajmabadi** [comment]
> Plus: You will have a new type in your assembly and public interface even though you just wanted to have a wrapper with some more meaning.

That sounds like a virtue.  You're defining new operations and whatnot.  how is exposing that as a "typedef" better than just exposing as a type?  Why do you need a new concept for this when this is a primary use case for one of hte existing concepts the system already provides?

**Comment by jnm2** [comment]
@Danielku15 

> > Meaning, you want to be able to pass your typedef type to things expecting ISomeInterface or SomeBaseClass
> 
> Yes, you can basically wrap anything in the typedef, a class, interface, enum or struct (by this also standard data types like string). Just like in C++ the typedef becomes an alias for the original type.

You can match this behavior today with a wrapper type using implicit conversions, and for interfaces, delegating implementations. The only thing this doesn't get you is reference equality with the wrapped value, if, say, you pass the wrapper to an `object` argument.

**Comment by Danielku15** [comment]
> That sounds like a virtue. You're defining new operations and whatnot. how is exposing that as a "typedef" better than just exposing as a type? 

The thing is rather that you are **not** exposing the typedef. It stays assembly-internal during compile time only the methods might remain as extension methods to the underlying type. 

> Why do you need a new concept for this when this is a primary use case for one of hte existing concepts the system already provides?

I wasn't aware that the runtime is supposed to be that intelligent when it comes to simple wrapper structs. My worries in this case are that everywhere, where you pass on your struct, it is cloned even though it might not be needed. I work since years on various cross platform projects and used Haxe quite a lot. I always loved how you can write platform specific wrappers without causing any overhead on your library footprint. When I moved to C# cross platform compilation I was always missing such a feature a bit. Especially when cross compiling to C# to JavaScript a small footprint and inlining are an important factor. 

But if the .net Runtime is very performant on this it might not be needed from this perspective.

**Comment by CyrusNajmabadi** [comment]
> I wasn't aware that the runtime is supposed to be that intelligent when it comes to simple wrapper structs. My worries in this case are that everywhere, where you pass on your struct, it is cloned even though it might not be needed. 

This is only a concern if you have a large struct.  i.e. a struct with many members in it.  In this case, this is about a struct always wrapping exactly one value.  In that case, the size of hte struct is **exactly** the same as the size of the underlying value, so the cost stays the same.  :)

**Comment by chrisrollins** [comment]
I continue to run into situations where I want this feature.

For example I have some methods that take string arguments, and then I thought of a reason to overload these methods a bunch of times because I wish to provide a lot of flexibility in my API. At some point I noticed that I had put the parameters in the wrong order for a few of the overloads. This is something I had to find out by either looking at the code carefully or by running into bugs during testing because the parameters share the same type.

My solution to avoid having this happen again is to define some structs that represent parameter types that the public overloads will actually pass to the implementation so that if I put them in the wrong order it will cause a compilation error.

I believe that there is runtime overhead to my approach and that bothers me. With typedef there would definitely be no runtime overhead in this situation, but I would still be able to use it to gain some type safety in the same way simply by using it to alias types like string as my parameter types instead of boxing them.

Another advantage of typedef is that it's more lightweight. In my case I had to define structs and give them public constructors even though all they really do is box a string value.

**Comment by louthy** [comment]
Related https://github.com/dotnet/csharplang/issues/410

**Comment by theunrepentantgeek** [comment]
@chrisrollins 

> My solution ... is to define some structs that represent parameter types that the public overloads will actually pass to the implementation so that if I put them in the wrong order it will cause a compilation error.

I do this all the time - and I find that these *semantic types* (usually value types implemented with **struct** instead of **class**) tend to become pervasive through the codebase, representing key concepts.

> I believe that there is runtime overhead to my approach and that bothers me. 

My experience is that unless you're talking about a method that's called thousands or millions of times in a tight loop, the overhead won't be significant nor perceptible. If you have concerns, it's worth running a proper benchmark to prove or refute things. I did one this week using [BenchMarkDotNet](https://benchmarkdotnet.org/Overview.htm) and discovered that the difference in my particular situation (which wasn't struct vs string) was a handful of *nanoseconds*.

**Comment by chrisrollins** [comment]
@theunrepentantgeek you are right. It won't matter unless the code is used at scale. But that's the only time it ever matters for performance. To me it's not acceptable that achieving type safety can make code unsuitable for scaling. If none of your code is being used at scale then performance doesn't matter anyway.

And keep in mind that isolated tests with very little happening does not reveal the true cost of boxing. You have more heap pointers so you have more cache misses. Of course it looks like nothing when it's isolated. That's not representative of the potential cost.

**Comment by CyrusNajmabadi** [comment]
> @theunrepentantgeek you are right. It won't matter unless the code is used at scale. 

Do we actually know that?  Is there actually a performance problem to address here with the struct approach?  Is that problem something that could be easily solved by the CLR, thus makignt his a perfectly fine solution for your needs, without any language change whatsoever?

**Comment by chrisrollins** [comment]
@CyrusNajmabadi yes more heap pointers = worse performance. The CPU cache is burdened more heavily, causing more cache misses. But yeah maybe they could just improve it under the hood, idk.

Even then, would you say it's not a worthy feature? At the very least it would make it quicker and more lightweight to define these types.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi yes more heap pointers = worse performance

Why?  If i wrap with a struct, i don't see why the runtime coudln't be smart enough to understnd that and treat that wrapper as effectively a no-op that just gets erased as far as passing/allocating/calling functions was concerned.

**Comment by CyrusNajmabadi** [comment]
> Even then, would you say it's not a worthy feature? At the very least it would make it quicker and more lightweight to define these types.

I don't think it's particularly slow or heavy to just create a wrapper struct :)   So i'm not really big on making it that much faster...

**Comment by chrisrollins** [comment]
@CyrusNajmabadi I am not sure why but various people have told me that it always boxes values inside of structs.

But anyway simple structs feel a bit smelly to me because I have to define a constructor to enable inline initialization. Although sometimes it's beneficial to include some logic in the constructor, #410 provides a really concise way to do a lot of things you'd be doing in the constructor. I really like that proposal even more than this one.

**Comment by theunrepentantgeek** [comment]
@chrisrollins 
> To me it's not acceptable that achieving type safety can make code unsuitable for scaling.

My experience is that type safety usually results in code that scales *better*. For example, it might enable better runtime optimization, or it might reduce code duplication, or it might permit better algorithms.

Actually, that's a good point - I've invariably found that meaningful performance gains come from better algorithms, not micro-optimizations.

@CyrusNajmabadi 

> Is there actually a performance problem to address here with the struct approach?

I didn't think there was. But, facts always trump opinions, so I decided to write a benchmark to find out. (BTW, I suspect this was a rhetorical question on your part as you almost certainly already know the answer.)

I created a simple project using [BenchMarkDotNet](https://github.com/dotnet/BenchmarkDotNet) and made sure to include benchmarking of allocations as well as execution time. [Source code](https://gist.github.com/theunrepentantgeek/9861b043f31d440b8e7183d14a347cc7)

To keep things simple, the three different approaches all just call `GetHashCode()` for the same string (I originally used `Length` the but the optimizer is just a little too clever and it managed to optimize away one of the methods completely.)

Here are the results:

Method                   |     Mean |        Error |      StdDev |   Median | Allocated |
--------------------- |----------:|-----------:|------------:|---------:|----------:|
Primitive                  | 6.092 ns | 0.2577 ns | 0.5437 ns | 5.875 ns |       0 B |
UsingStruct             | 5.916 ns | 0.1123 ns | 0.1050 ns | 5.904 ns |       0 B |
UsingGenericStruct | 5.846 ns | 0.0706 ns | 0.0660 ns | 5.829 ns |       0 B |

I did all of this on my laptop, and while I was "hands off" while the benchmark was running, this isn't a machine that's been configured to eliminate as much background activity as possible. The only tuning I did was to configure power management for *best performance*.

Given the size of the error bars, my conclusion between these techniques is that there's no meaningful difference in performance. Crucially, they all run with **zero memory allocation**.

@chrisrollins I've published [my source as a gist](https://gist.github.com/theunrepentantgeek/9861b043f31d440b8e7183d14a347cc7) so that you can grab it and play around with the code to draw your own conclusions.

**Comment by chrisrollins** [comment]
Ok perhaps I heard wrong and it does optimize away structs in certain cases.

I still think it's a good feature for making code more concise.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi I am not sure why but various people have told me that it always boxes values inside of structs.

That would be very bad :)

**Comment by jnm2** [comment]
@chrisrollins It's the other way around: storing any value type as an `object` or an interface type requires boxing, meaning the data is copied from its current location to a new location on the heap in order for there to be something for the object/interface instance to be referencing.

**Comment by chrisrollins** [comment]
@jnm2 that makes sense. thanks.

**Comment by LeonG-ZA** [comment]
This is a feature that I absolutely hated in C/C++. I would go through code and scratch my head just to figure out...oh this is just [some common type].
It *is* useful. In C/C++ it is useful for typedefing different types based on the architecture you are building on, but it often gets misused A LOT. 
When I quickly browse through new code I would rather want to see the real full names of types I am familiar with.

**Comment by bondsbw** [comment]
IntelliSense can help.  Hovering over a type identifier should show the resolved type in Quick Info.

Of course the feature could be abused.  Someone could replace all the type names in their code base with acronyms or senseless names.  Don't do that.

**Comment by LeonG-ZA** [comment]
Yes, it does show up. A lot of times when I quickly scroll through new code I can get a general idea of what some class is doing, but I do have to occasionally stop. Usually it is when I encounter these:
- var someVariable = someStuff...;  //I have to hover over var or someStuff... to figure out the type.
- I see some unknown method being used only to find out  'using static' was used so the class is hidden through a quick glance.

> Of course the feature could be abused. Someone could replace all the type names in their code base with acronyms or senseless names. Don't do that.

True, I wouldn't of course. Some people just see some feature as a silver bullet that they always have to use. We once had a new guy working for us who used Tuples for EVERYTHING he touched. The code was a nightmare and had to be scrapped/rewritten. Also our fault because we didn't do code reviews at the time.

**Comment by bondsbw** [comment]
I suspect more was at play if the entire codebase had to be rewritten.  It's straightforward to create a named type that has the same fields as a tuple.

I wouldn't be surprised if a refactoring exists to do this for you.

**Comment by jnm2** [comment]
@CyrusNajmabadi ^

**Comment by CyrusNajmabadi** [comment]
The trick is to ```@```at me three times.  Then i'll write the feature for you (and also possibly haunt you).

**Comment by CyrusNajmabadi** [comment]
Note: the only thing holding me back on this feature is that i need to sit down and think what it means to replace a tuple-type with a named type.  Should that fix up all complimentary tuple types across your codebase?  What if you do this:

```c#
(int, int) GetPair()
{
    //...
    return (x: 0, y: 1); // <- intro named type here.
}
```

What happens in a case like that?

Or, what if someone just uses ``(int x, int y)``` elsewhere (like even an unrelated project!), should that get updated.  Should we try to just walk the usages here and only update code that is actually affected?  etc. etc.

Basically, it's non-trivial to figure out what code actually needs fixing up.  this is why my recent work has stuck with just converting anonymous-types.  It's very clear where you can stop looking when using those guys.

**Comment by kofifus** [comment]
Sorry if this may be a little bit off topic ..

The OP is about "Proposal: Add support for typedef (like in C++) ", typedef in C++ are just aliases-  that is they do not create new types .. for example a function accepting a parameter of type T will compile when called with any T typedef, this is equivalent to C# using statement and C++11 using statement.

If this is what the OP is after I think the 'internal using' suggestion https://github.com/dotnet/csharplang/issues/1300#issuecomment-364228173 is a perfect solution.

However, IMO what is more useful is having a way to define a _new type_ which is a duplicate of an existing type, I feel this is essential for strongly typed languages like C#

Consider `void f(string name, string address) => ...`

This is an inherently dangerous function, it invites bugs like `f("1 Victoria St", "John")` which are easy to create, deadly, and hard to find  ... what you really want is a way of saying something like:

```
using Name = newtype(System.String)
using Address = newtype(System.String)
void f(Name name, Address address);
```

and have Name and Address as distinct types so that 
```
var name = new Name("John");
var address = new Address("1 Victoria St");
f(address, name); 
```

will not compile. These distinct type can then be used in function resolution, generic parameters etc ...

Basically this is what C++ with Boost achieves with BOOST_STRONG_TYPEDEF 
 
ATM the only way to achieve this in C# is through composition which means proxying all methods and can be extremely cumbersome. Maybe I should start a new issue for this ?

**Comment by aaronfranke** [comment]
@kofifus I agree, that would be a useful feature, but also it is indeed different from what I had in mind.

If the goal is solely for type safety purposes with data storage, a `struct` wrapper could work for that, but would be overkill and impractical for many use cases, compared to `using` currently.

**Comment by theunrepentantgeek** [comment]
@kofifus I think a separate issue would be a good idea.

I'm a [big](http://www.nichesoftware.co.nz/2015/09/27/semantic-types.html) [fan](http://www.nichesoftware.co.nz/2016/05/27/semantic-types-in-csharp6.html) of semantic types (those links are to blogs I've written on the topic), but I have real concerns with this idea.

The `String` type has 2 methods and 43 methods, not counting overrides or extension methods; how many of them are actually relevant to a `Name` or `Address` type?

Similarly, if you defined `CustomerId = newtype(System.Int32)`, that class would acquire 8 methods - including `TryParse()`, supporting values like `0xDEADBEEF`.

I strongly believe that semantic types should choose what methods they expose, not acquire them via bulk dump.

**Comment by kofifus** [comment]
theunrepentantg - you're perfectly write about that but I chose a very simple case to demonstrate ...

I'm more in need of things like
`using TheDict = newtype(ImmutableDictionary<Address, Tuple<Address, Name>>);`

I believe that if defining types was a very cheap one line operation it will make them extremely usable ..

**Comment by jzabroski** [comment]
> You can do aliasing currently for types based on defines with use `using ... = ...` with `#if` to achieve the `float`/`double`

I think the drawback here is that the portability/discoverability of the API consuming code is reduced.  `#if` is not late-bound, and API consumers cannot "know" about the dynamic targeting of "real" - if they decompile the nuget package, they will see whatever implementation the package chose, not the source code allows.

On Android, enums can be defined using a special TypeDef annotation syntax, which effectively allows enums to be treated as inline constants.  In fact, this seems to be the preferred way to do things to save MB due to the horrendous size allocation for Java enums.

**Comment by filimonic** [comment]
This would be nice to separate types of one type and different representation.
Example - GPS Coordinates - Latitude and Longitude. 
1. I want separate them apart not to swap them occasionaly. Using typedef is good here.
2. When I want to switch them from double to decimal, I need to change only two lines of typedefs.

**Comment by dozer75** [comment]
`typedef` or extending `using` could be nice with tuples too to reduce code when using tuples:

`typedef MyTuple (int Id, string Name, DateTime Created)` or `using MyTuple = (int Id, string Name, DateTime Created)`

Could result in more readable code (example simplified):

```
private MyTuple GetInfo(int id)
{
    return something.Select(item => (item.Id, item.Name, item.Created)).FirstOrDefault();
}

private void PrintTuple(MyTuple info)
{
    console.WriteLine(info.Id);
    console.WriteLine(info.Name);
    console.WriteLine(info.Created);
}
```

Of course you can create a class or struct, but that's more code and in some circumstances (like in internal code) it seems more reasonable to simplify it using a tuple instead of a class or struct.

**Comment by aaronfranke** [comment]
Here's another thought: However this feature is implemented, it doesn't really need to be global, as long as it allows spanning many files. Perhaps a namespace-level `using` directive which includes that `using` statement for all code in the same namespace, and maybe all files including that namespace?

This would be a more generic approach and could also allow things like `using static` to exist everywhere in the namespace.

**Comment by Ozzard** [comment]
I have two use cases for this, both heavily performance-related:

* I have a simulation application that represents money, where the same codebase sometimes needs decimal for full precision, sometimes double for higher speed and damn the inaccuracies.  "using Money=double" doesn't scale across high tens of files.
* https://github.com/PaulBrack/Yamato where we're doing huge amounts of storage and calculation on mass spectrometry files - it's not unusual for the calculations on a large file to need a couple of hundred gigabytes of working memory.  Again, doubles for precision, or floats for reducing the storage requirement and being "mostly right", which is very useful in some cases.

Both of these could be mostly solved by allowing generics to take a restriction of "where T: Number" or similar (list the operations that must be supported, perhaps).  Unfortunately this restriction is very difficult to enforce on the primitive types, as for complex reasons they're not first-class citizens in the type space.

**Comment by CyrusNajmabadi** [comment]
@Ozzard 

`struct Money { private readonly double _value; }` ?

It allows you to ensure you're not innappropriately mixing incorrect types.  And, nicely, the runtime will effectively elide everything and you should have full double-perf.

**Comment by Trayani** [comment]
@CyrusNajmabadi ,
Out of sheer curiosity (and uncertainty), does it mean that the struct is treated as double without the indirection once it gets compiled? If so, it is pretty cool. How is this optimization called? I would like to read more on it.
(sorry for the off-topic)

**Comment by CyrusNajmabadi** [comment]
> Out of sheer curiosity (and uncertainty), does it mean that the struct is treated as double without the indirection once it gets compiled?

Effectively yes.  It will still be in the metadata signatures as Money, but that's a good thing.

> How is this optimization called? 

It's just good inlining.  The runtime will see that all your members are very simple, and this struct is just a simple wrapper around double.  So effectively all jitting will elide everything unnecessary.

**Comment by ericwj** [comment]
Much of this can be done with some `#define` lines added to files copied from some primary file that you edit by hand. Like @Ozzard's use cases. If you don't write `decimal` or `double` but `number` with at the top of the file something like:

```
#if USE_DECIMAL
using number = System.Decimal; // use as decimal
#elif USE_DOUBLE
using number = System.Double; // use as double
#else
using number = System.Single; // use as float
#endif
```
So you will see the default in code files as `float` since there are no defines in the file.

Then you can generate a new file which contains e.g. `#define USE_DECIMAL` and after that a copy of the file you were editing.

There is a new feature in the newest C# in preview with which this can be made much cleaner than what it gets to look like as described here, but even without code generation this can pretty easily be done with pre-build events (`copy a.cs + b.cs c.cs`) or in MSBuild or even in a separate project as a custom build step.

**Comment by aaronfranke** [comment]
@ericwj The problem right now is that this definition has to be repeated in every file. There should be a project-level feature to include such kinds of definitions for usage in every file without having to copy it into each file and without setting up complicated build events (and if I'm understanding this, "number" would not show up as a valid type with intellisense if adding it was a pre-build event).

**Comment by HaloFour** [comment]
This has been championed, which means that the team will at least discuss it: https://github.com/dotnet/csharplang/issues/3428

**Comment by ericwj** [comment]
It will work with IntelliSense regardless of how you produce the file. It will just be plain C# that is part of the normal compilation.

This might be useful for some people because 
* you don't need to sit on your hands while the Gods ('at'MadsTorgersen e.a.) might or might not favor this.
* having this baked in `typedef` will give worlds of hurt with C++-like project configuration hell and lots of build configurations which .NET developers usually have the luxury of ignoring completely
* as opposed to `typedef` like in C++ this in fact allows the variants produced as described to live in the same project and be used in different places at the same time without recompiling

**Comment by osamakawish** [comment]
Regarding the issue that it has to be repeated in every file, can this not be allowed to be included in the namespace directly? To avoid the repetition?

**Reply by keithn** [reply]
I'd like a syntax similar to F#s way of defining types ```type X = Y```,  I think it should also cover discriminated unions also, though that seems to be trying to use enum, as per https://github.com/dotnet/csharplang/issues/113       Semantic types are a real pain in the ass for C# when they  simply have an equivalence with a primitive type.

**Comment by badcel** [comment]
My usecase is wrapping a C library which uses `typedef` in its own code to rename primitive value types. To keep the type as expressive as in C and not fall back to the primitive type if writing the pinvoke call it would be nice to have a typedef in C# which is like a global alias for a given type avoiding wrapping it into a struct. Thus matching the C API as close as possible.

**Reply by bernd5** [reply]
You can write in a single file, e.g. "Usings.cs":

```csharp
global using Foo = System.Int32;
```

Then you can use it:
```csharp
Foo x = 12;
```

**Reply by badcel** [reply]
It looks like the global using is not accessible outside of my library? If I’m wrapping a C library to make it available in the C# world users of my library would still see `int` instead of `Foo`.

**Comment by artdedeco** [comment]
I'd like to have this and have the "typedef" exported. My use case: export "default" generic type instantiation without inheritance:

```cs
export using Foo = Foo<Default>;

public struct Foo<T>
{
}
```

I have used this kind of pattern many times for different reasons (making API consumption simpler for the typical use case being the most common, but also for creating testable code with DI through generics). However, today was the first time I wanted to use this for a struct, and I can't use inheritance there.

**Comment by svew** [comment]
I keep wishing for a thing like this, it would make my Rest API requests much easier to write with strict typing:

```cs
interface IRequest
{
    typedef Response;
}

record MyRequest
{
    using Response = MyResponse
}

record MyResponse
{
}

Task<T.Response> Send<T>(T request)
    where T : IRequest
{
    HttpResponseMessage responseMsg = // send the request...
    return JsonConvert.DeserializeObject<TRequest.Response>(await responseMsg.Content.ReadAsStringAsync());
}

// Call it like this:
MyResponse response = await Send(new MyRequest());
```

I can do to this without "typedef" right now using abstract classes, but once you start making combinations of different request content serializations (json, binary, no body) and response serializations (json, binary, no body), things start getting *very* messy...

**Reply by tannergooding** [reply]
As was covered above, `global using alias = Type;` works and makes it accessible to every file in the project.

You can even define these at the msbuild (csproj/props/targets) level, which means that you can include it in a NuGet package or similar.

There is nothing blocking or lacking here and I doubt the language will do anything to make it even "simpler", such as embedding such metadata in an assembly. The existing tools are more than sufficient for the niche edge cases you need it.

**Reply by svew** [reply]
@tannergooding I don't believe global aliasing addresses what I'm talking about. I'm looking for is the ability to add types as members of classes/structs, in order to associate an otherwise unrelated type to it. C++ uses that to allow users to [create type overrides for things such as allocators](https://en.cppreference.com/w/cpp/memory/allocator_traits.html). I'm looking for something far simpler, but still similar, in the ability to define types (e.g. `Response`) as members of classes/structs (e.g. `using Response = MyResponse`) so that I can create a clear type relationship between an HTTP request's JSON body and the expected JSON response.

I don't believe there's a way to do this right now, except for creating a subclass of a generic superclass, but as mentioned doing that sort of thing when there's many types to define gets messy

**Reply by svew** [reply]
I found a more relevant thread for my interest: https://github.com/dotnet/csharplang/issues/8712

**Reply by tannergooding** [reply]
Associated types would be a different feature with its own discussion thread.

This discussion thread was for what is effectively `typedefs`, which has always existing in the language (per file at least) via `aliases` and has had the ability to define per project for several releases at this point.

**Comment by aaronfranke** [comment, marked as answer]
The new `global using` feature resolves this, so I'll close this discussion.

**Reply by jzabroski** [reply]
Yes, but it would also be nice if it could actually define new types rather an alias to an existing identity. That should be a new discussion thread.

**Reply by tannergooding** [reply]
Defining new strong types can effectively be accomplished with records and a couple helper operators. Source generators are a viable option as well.

Anything further would require its own discussion thread, which I'm sure may already exist.

**Reply by miyu** [reply]
@tannergooding To clarify, if one is looking for typedef-like behavior but believes `global using `does not address their use-case, should they open another issue or continue responding to this one, given this discussion is marked as closed but other discussions are pointing to it as duplicates?

In my case, I'd like a Source Generator to expose private typedefs within a class, e.g.

```csharp
[UseSomeSourceGenerator]
public partial class SomeClass<T, U, V> {
   private void SomeMethod1(TArg arg) { }
   private void SomeMethod2(TArg arg) { }
}
```

And via the source generator:

```csharp
// via [UseSomeSourceGenerator]
public partial class SomeClass<T, U, V> {
    private alias TArg = A<T, B<U, V>, C<T, U, D>>;  
    private class D { }
}
```

SomeMethod would compile to have arg A<T, B<U, V>, C<T, U, D>>, potentially with metadata indicating the SomeSourceGeneratedType alias. Other types could be defined which alias `TArg` to something else within their scope.

This isn't solved by `global using` without significant boilerplate and (to my understanding) making D public. Similarly, one could imagine a case where `TArg` could be public, but many types would define their own `TArg` as is quite common with C++. 

In my case, I frequently have lengthy mutually-dependent CRTP types  (https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern) - For example, situations where N=3-4 structs refer to each other in type args and each of the N structs has M possible variants. In such a situation, global aliases typically do not make sense (as your alias count is combinatoric, N * N^M), whereas within specific types where you know the specific concrete types used for each of the N types. Such patterns are common in high-perf code.

**Reply by jzabroski** [reply]
File-based aliases should work for that. Just make sure there is one file per source generated combination. You can store the combination resolutions via the file name e.g. SomeClass gets specialized to
```
StockQuote`string`DateTime`int`double.cs
```

Although, you can also just used named tuples.

**Reply by ericwj** [reply]
Would be nice if the compiler would remember the alias and show it in IntelliSense and hovers, since even though they are mere aliases, they have meaning and there is a reason for their use.

For example, the alias might be different depending on `#define`s and hence the file they are in and it is highly unproductive to see them being e.g. naked `int`, no longer knowing why and unable to reason whether that is even correct without browsing to definition manually.
